opacus/grad_sample/grad_sample_module.py

```diff
@@ -260,6 +298,94 @@ def is_supported(cls, module: nn.Module) -> bool:
         """Check if this module is supported"""
         return type(module) in cls.GRAD_SAMPLERS or type(module) is DPLSTM
 
+    def ddp_backward_callback(self, engine, p, grad):
+        """
+        This hook operates like PrivacyEngine.step(), but on a single layer:
+        1. clip_and_accumulate
+        2. get the clip_values with clipper.pre_step()
+        3. add the noise
+        """
+
+        # Get the norm of the gradient for each sample for this layer
+        all_norms = calc_sample_norms(
+            named_params=((None, p.grad_sample),),
+            flat=False,
+        )
+
+        # Get the constant clipping factor for a single layer
+        clipping_factor = engine.clipper.norm_clipper.calc_clipping_factors(all_norms)
+        assert len(clipping_factor) == 1
+        clip_factor = clipping_factor[0]
+        batch_size = p.grad_sample.shape[0]
+
+        # Do the clipping
+        summed_grad = engine.clipper._weighted_sum(clip_factor, p.grad_sample)
+
+        # accumulate the summed gradient for this mini-batch
+        if hasattr(p, "summed_grad"):
+            p.summed_grad += summed_grad
+        else:
+            p.summed_grad = summed_grad
+
+        del p.grad_sample
+
+        # Average (or sum) across the batch
+        # TODO: add support for Poisson batch sampling
+        res = engine.clipper._scale_summed_grad(p.summed_grad, batch_size)
+
+        del p.summed_grad
+
+        # NOTE: `self.clipper.pre_step()` returns a tensor with n=n_params clip_values
+        # where each clip_value is the L2 norm of `threshs`:
+        #   `max_norm = threshs.new_full((n,), threshs.norm(2))`
+        # With per layer clipping, `threshs` is a tensor with n flat_values
+        assert len(engine.clipper.norm_clipper.thresholds) == 1
+        clipping_thresh = engine.clipper.norm_clipper.thresholds[0]
+        clip_value = (self.n_params ** 0.5) * clipping_thresh
+        # if engine.rank == 0:
+        #     print(
+        #         f"max_grad_norm: {engine.max_grad_norm} \nclipping_tresh: {clipping_thresh}\nclip_value: {clip_value}"
+        #     )
+
+        noise = _generate_noise_ddp(engine, clip_value, res)
+        if engine.loss_reduction == "mean":
+            noise /= batch_size
+
+        # Only one GPU adds noise
+        if engine.rank == 0:
+            res += noise
+
+        return res
+
+
+def _generate_noise_ddp(
+    engine, max_grad_norm: float, reference: nn.parameter.Parameter
+) -> torch.Tensor:
```

Comment:
Point to discuss.
Is there any way to decouple this from the core GradSampleModule? The idea behind recent refactoring, which introduced GradSampleModule, was exactly to avoid the coupling between per sample gradient computation, clipping and noise generation.
This change seems to reverse this. Although I understand that doing clipping & noise within the backward hook is the very core idea of this change, so it might be tricky

Reply:
Yes, it would be neater to decouple the clipping/noising. I thought about it, and I can try to divide this piece into two parts:
- the per-layer clipping/noising hooks (`ddp_backward_callback` and utils such as `_generate_noise_ddp`) would be defined in PrivacyEngine, and called by the hook helpers
- the hook helpers (add/remove ddp hook) would stay in GradSampleModule

In addition to being neater, this structure might make easier to modify the DDP functionality (e.g., per-layer thresholds) without having to change GradSampleModule. @ffuuugor what do you think?

Another option would be to pack the DDP functions into a separate module, similarly to `dp_ddp.py`

Reply:
Oops, sorry for the delay
I like the idea of moving DP-specific functionality (e.g. `_generate_noise_ddp`) to PrivacyEngine. Do I understand correctly, that you'll implement the hook that does noise and clipping somewhere else, and will pass it to `add_ddp_hook` in GradSampleModule?

The principle we want to adhere to is GradSampleModule should make sense outside of DP context as well

Reply:
Exactly -- with a small change: when I implemented this design, I found it more convenient to define and register the hooks directly from the PrivacyEngine. I still give a list of handles to GradSampleModule, to make it easier to remove the hooks later.

I think that I can go even further and completely leave GradSampleModule untouched, if I implement `remove_ddp_hooks` into the PrivacyEngine directly. @ffuuugor please let me know if you would like me to do that, after you've had a look at the new code